mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h

```diff
@@ -21,42 +21,224 @@ class SparseTensorLevel {
+  ///
+  /// For dense level, the *posLo* is the linearized position at beginning,
+  /// while *loopHi* is the largest *coordinate*, it also implies that the
+  /// smallest *coordinate* to start the loop is 0.
+  ///
+  /// For sparse level, [posLo, loopHi) specifies the range of index pointer to
+  /// load coordinate from the coordinate buffer.
+  ///
   /// `bound` is only used when the level is `non-unique` and deduplication is
   /// required. It specifies the max upper bound of the non-unique segment.
   virtual std::pair<Value, Value> peekRangeAt(OpBuilder &b, Location l, Value p,
-                                              Value bound = Value()) const = 0;
+                                              Value segHi = Value()) const = 0;
 
+  Level getLevel() const { return lvl; }
   LevelType getLT() const { return lt; }
-  Value getPos() const { return pos; }
-  Value getCrd() const { return crd; }
-  Value getLoopHi() const { return loopHi; }
-  Value getLoopLo() const { return loopLo; }
+  Value size() const { return lvlSize; }
+
+  //
+  // Level properties
+  //
+  bool isUnique() const { return isUniqueLT(lt); }
 
 protected:
-  SparseTensorLevel(LevelType lt, Value lvlSize)
-      : lt(lt), lvlSize(lvlSize), pos(nullptr), crd(nullptr), loopHi(nullptr),
-        loopLo(nullptr){};
+  SparseTensorLevel(unsigned tid, unsigned lvl, LevelType lt, Value lvlSize)
+      : tid(tid), lvl(lvl), lt(lt), lvlSize(lvlSize){};
 
+public:
+  const unsigned tid, lvl;
   const LevelType lt;
   const Value lvlSize;
+};
+
+enum class IterKind : uint8_t {
+  kTrivial,
+  kDedup,
+  kSubSect,
+  kNonEmptySubSect,
+  kFilter,
+};
+
+/// Helper class that helps generating loop conditions, etc, to traverse a
+/// sparse tensor level.
+class SparseIterator {
+  SparseIterator(SparseIterator &&) = delete;
+  SparseIterator(const SparseIterator &) = delete;
+  SparseIterator &operator=(SparseIterator &&) = delete;
+  SparseIterator &operator=(const SparseIterator &) = delete;
+
+protected:
+  SparseIterator(IterKind kind, unsigned tid, unsigned lvl,
+                 MutableArrayRef<Value> itVals)
+      : kind(kind), tid(tid), lvl(lvl), crd(nullptr), itVals(itVals){};
+
+  SparseIterator(IterKind kind, const SparseIterator &wrap)
+      : kind(kind), tid(wrap.tid), lvl(wrap.lvl), crd(nullptr),
+        itVals(wrap.itVals){};
+
+public:
+  virtual ~SparseIterator() = default;
+
+  Value getCrd() const { return crd; }
+
+  ValueRange getItVals() const { return itVals; };
+  void seek(ValueRange vals) {
+    assert(vals.size() == itVals.size());
+    std::copy(vals.begin(), vals.end(), itVals.begin());
+    // Now that the iterator is re-positioned, the coordinate becomes invalid.
+    crd = nullptr;
+  }
+
+  //
+  // Iterator properties.
+  //
+
+  // Whether the iterator support random access (i.e., support look up by
+  // *coordinate*).
+  // A random access iterator also traverses a dense space.
+  virtual bool randomAccessible() const = 0;
```

Comment:
Although I am a huge fan of very compact code, I would use vertical space between the different methods here, since we are describing the interface here